dpc/src/state/state.rs

```diff
@@ -0,0 +1,122 @@
+    outputs: Vec<Output<C>>,
+    joint_serial_numbers: Vec<u8>,
+    errors: Vec<String>,
+}
+
+impl<C: Parameters> StateBuilder<C> {
+    /// Initializes a new instance of `StateBuilder`.
+    pub fn new() -> Self {
+        Self {
+            inputs: Vec::new(),
+            outputs: Vec::new(),
+            joint_serial_numbers: Vec::new(),
+            errors: Vec::new(),
+        }
+    }
+
+    /// Adds the given input into the builder.
+    pub fn add_input(mut self, input: Input<C>) -> Self {
+        // Ensure there are no outputs assigned yet, as the builder computes joint serial numbers.
+        if !self.outputs.is_empty() {
+            self.errors.push("Builder cannot add new inputs after outputs".into());
+        }
+
+        match self.inputs.len() < C::NUM_INPUT_RECORDS {
+            true => self.inputs.push(input),
+            false => self.errors.push("Builder exceeded maximum inputs".into()),
+        };
+        self
+    }
+
+    /// Adds the given output into the builder.
+    pub fn add_output(mut self, output: Output<C>) -> Self {
+        match self.outputs.len() < C::NUM_OUTPUT_RECORDS {
+            true => self.outputs.push(output),
+            false => self.errors.push("Builder exceeded maximum outputs".into()),
+        };
+        self
+    }
+
+    /// TODO (howardwu): TEMPORARY - `noop: Arc<NoopProgram<C>>` will be removed when `DPC::setup` and `DPC::load` are refactored.
+    /// Finalizes the builder and returns a new instance of `State`.
+    pub fn build<R: Rng + CryptoRng>(&mut self, noop: Arc<NoopProgram<C>>, rng: &mut R) -> Result<State<C>> {
+        // Ensure there are no errors in the build process yet.
+        if !self.errors.is_empty() {
+            for error in &self.errors {
+                eprintln!("{}", error);
+            }
+            return Err(anyhow!("State builder encountered build errors: {:?}", self.errors));
+        }
+
+        // Construct the inputs.
+        let mut inputs = self.inputs.clone();
+        // Pad the inputs with noop inputs if necessary.
+        while inputs.len() < C::NUM_INPUT_RECORDS {
+            // TODO (howardwu): Decide whether to "push" or "push_front" for program flow.
+            inputs.push(Input::new_noop(noop.clone(), rng)?);
+        }
+
+        // Update the joint serial numbers.
+        self.update_joint_serial_numbers()?;
+
+        // Construct the outputs.
+        let mut outputs = self.outputs.clone();
+        // Pad the outputs with noop outputs if necessary.
+        while outputs.len() < C::NUM_OUTPUT_RECORDS {
+            let position = (C::NUM_INPUT_RECORDS + outputs.len()) as u8;
+            // TODO (howardwu): Decide whether to "push" or "push_front" for program flow.
+            outputs.push(Output::new_noop(noop.clone(), position, rng)?);
+        }
+
+        Ok(State { inputs, outputs })
+    }
+
+    fn update_joint_serial_numbers(&mut self) -> Result<()> {
+        // Compute the joint serial numbers.
```

Comment:
I feel maybe we can add an `assert_eq` that `self.inputs` is exactly `C::NUM_INPUT_RECORDS`. This is to ensure that the `joint_serial_numbers` always include the necessary padding of noop inputs.

Reply:
<- Needs confirmation: is that the expected behavior?

Reply:
Yes, that is the intended behavior.

I will add a check there.

Reply:
I'll also mention based on how the struct is designed, there shouldn't be a case whereby more than `C::NUM_INPUT_RECORDS` will be admitted into the struct

Reply:
As such, I will make the check a `debug_assert_eq!` if that is alright with you

Reply:
Done. Will add tests in the next PR to ensure behavior of `State` struct matches basic expectations